src/net.cpp

```diff
@@ -1939,14 +1941,23 @@ void CConnman::ThreadOpenConnections(const std::vector<std::string> connect)
                 LogPrint(BCLog::NET, "Making feeler connection to %s\n", addrConnect.ToString());
             }
 
-            // Open this connection as block-relay-only if we're already at our
-            // full-relay capacity, but not yet at our block-relay peer limit.
-            // (It should not be possible for fFeeler to be set if we're not
-            // also at our block-relay peer limit, but check against that as
-            // well for sanity.)
-            bool block_relay_only = nOutboundBlockRelay < m_max_outbound_block_relay && !fFeeler && nOutboundFullRelay >= m_max_outbound_full_relay;
+            ConnectionType conn_type;
+            // Determine what type of connection to open. Open OUTBOUND
+            // connections until we meet our full-relay capacity. Afterwards,
+            // open BLOCK_RELAY connections until we hit our block-relay peer
+            // limit.
+            if (fFeeler) {
+                conn_type = ConnectionType::FEELER;
+            } else if (nOutboundFullRelay < m_max_outbound_full_relay) {
+                conn_type = ConnectionType::OUTBOUND;
+            } else if (nOutboundBlockRelay < m_max_outbound_block_relay) {
+                conn_type = ConnectionType::BLOCK_RELAY;
+            } else {
+                //this case should be unreachable
+                conn_type = ConnectionType::OUTBOUND;
```

Comment:
totally agree the current state is weird. 

I dug more into this case & turns out.. I think the comment is wrong & the code is right. here's my understanding-

[this previous conditional](https://github.com/bitcoin/bitcoin/blob/master/src/net.cpp#L1868) earlier in the function means if full-relay & block-relay are both at capacity, one of the following happens: 
(A)`fFeeler` will be set to true 
(B) we continue to next iteration of the while loop or 
(C)`GetTryNewOutboundPeer` is true, `fFeeler` stays false, and we continue to the rest of the function. 

`GetTryNewOutboundPeer` is simply a wrapper for `m_try_another_outbound_peer`, which is set to true when a stale tip is detected ([link](https://github.com/bitcoin/bitcoin/blob/master/src/net_processing.cpp#L3817)).

so, when we get to this current conditional, if `fFeeler` if false & our full-relay and block-relay connection counts are at the allocated max -> `GetTryNewOutboundPeer` must be true. I believe the intent in this scenario is to open a new outbound connection.

--

This got me taking a closer look at `GetTryNewOutboundPeer` & I'm confused about how it works. 

It doesn't look like eviction / outbound counting logic checks for this bool. So if `m_try_another_outbound_peer` is true, the node would add an outbound connection, then `GetExtraOutboundCount` would count it as extra & `EvictExtraOutboundPeers` would evict a connection. 

Maybe this is a mechanism for peer rotation? Not sure how that would work though because evicting extra outbounds is [triggered](https://github.com/bitcoin/bitcoin/blob/master/src/net_processing.cpp#L1187) every 45 seconds, and decides what connection to evict based on the oldest recent-block-announcement. A newly added connection defaults to a value of 0 so would probably be selected for eviction? 

--

anyway, part 2 is a slight tangent from the question / review comment at hand. I'm going to delete this comment. if my part 1 logic makes sense, maybe I can add a comment thats actually true 😛 . 

but in the meanwhile, I'll be digging into understanding part 2 (and would appreciate any input!)

Reply:
Ok, I confirm what you write above and I verified that the new code behaves in an identical way with the old one, so this PR is good (except the `//this case should be unreachable` comment).

But the code (old and new) is confusing - I imagine somebody may attempt to "fix" this code which opens a connection even when both "full" and "block relay" capacities are exhausted.

What about checking here that `GetTryNewOutboundPeer()` is `true` instead of relying that if `fFeeler` is `false` then `GetTryNewOutboundPeer()` must have been true 85 lines above in the code?

It is kind of easier to understand that `GetTryNewOutboundPeer()` being `true` acts as an "override the limits and open a connection anyway".

Reply:
Without restructuring `ThreadOpenConnections` completely, perhaps we can just do a simple `assert(false)` here just so that if the previous conditional changes, we'll notice in test failures?

Reply:
@vasild - I looked into starting with the conditional of `GetTryNewOutboundPeer()`, but I think that subtly changes the logic of the function. on current master, block-relay-only takes precedence over GetTryNewOutboundPeer, so starting with that check would change the behavior slightly. I don't feel comfortable enough with all scenarios to evaluate if that change is fine, so here's what I did as a tiny improvement: [link](https://github.com/bitcoin/bitcoin/pull/19316/files#diff-9a82240fe7dfe86564178691cc57f2f1R1942). does this seem any better?

Reply:
Yes, it looks better to me (until the day when this assertion kicks in :-D)